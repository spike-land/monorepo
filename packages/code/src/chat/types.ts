import { ReactNode } from "react";

import { Message } from "../types/Message";
export type { Message };

export interface ChatHeaderProps {
  isDarkMode: boolean;
  toggleDarkMode: () => void;
  handleResetChat: () => void;
  onClose: () => void;
}

export interface ChatContainerProps {
  messages: Message[];
  editingMessageId: string | null;
  editInput: string;
  setEditInput: (value: string) => void;
  handleCancelEdit: () => void;
  handleSaveEdit: (messageId: string) => void;
  handleEditMessage: (id: string) => void;
  isStreaming: boolean;
  messagesEndRef: React.RefObject<HTMLDivElement>;
}

export interface MessageInputProps {
  input: string;
  setInput: (value: string) => void;
  handleSendMessage: (value: string, screenshot: string) => void;
  isStreaming: boolean;
  inputRef: React.RefObject<HTMLTextAreaElement>;
  isScreenshotLoading: boolean;
  screenshotImage: string | null;
  handleScreenshotClick: () => void;
  handleCancelScreenshot: () => void;
}

export interface ChatWindowProps {
  isOpen: boolean;
  children: ReactNode;
}
