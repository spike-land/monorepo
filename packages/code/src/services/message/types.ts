/**
 * Supported message types in the messaging system
 */
export enum MessageType {
  /** Regular text message */
  TEXT = "text",
  /** Command message for system operations */
  COMMAND = "command",
  /** Status update message */
  STATUS = "status",
  /** Error message */
  ERROR = "error",
}

/**
 * Role of the message sender
 */
export type MessageRole = "user" | "system" | "assistant";

// Re-export essential message types from lib interfaces
export type { ImageUrlPart, Message, MessageContent, TextPart } from "@/lib/interfaces";

/**
 * Response structure for message handling operations
 */
export interface MessageResponse {
  /** Indicates if the operation was successful */
  success: boolean;
  /** Optional data returned from successful operations */
  data?: unknown;
  /** Error message in case of failure */
  error?: string;
}

/**
 * Configuration options for the MessageHandler service
 */
export interface MessageHandlerConfig {
  /** Whether to log errors to console (default: true) */
  logErrors?: boolean;
  /** Maximum number of retries for failed operations (default: 3) */
  maxRetries?: number;
  /** Timeout in milliseconds for operations (default: 5000) */
  timeout?: number;
}
