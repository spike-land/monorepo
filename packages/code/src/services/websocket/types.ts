import type { RenderedApp } from "@/lib/interfaces";

export interface IWebSocketManager {
  init(): Promise<void>;
  cleanup(): void;
}

export interface IScreenshotService {
  takeScreenshot(): Promise<void>;
}

export interface IRenderService {
  updateRenderedApp(params: { transpiled: string; }): Promise<RenderedApp | null>;
  handleRender(renderedNew: RenderedApp | null): Promise<{ css: string; html: string; } | false>;
  cleanup(): void;
}

export interface IMessageHandlerService {
  handleMessage(event: MessageEvent): Promise<void>;
}

export interface IServiceWorkerManager {
  setup(): Promise<void>;
}
