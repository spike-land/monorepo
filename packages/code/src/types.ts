import type { Logger, LogLevel } from "./util";

/**
 * The mechanism to use for lazy-loading stylesheets.
 *
 * Note: <kbd>JS</kbd> indicates a strategy requiring JavaScript (falls back to `<noscript>` unless disabled).
 *
 * - **default:** Move stylesheet links to the end of the document and insert preload meta tags in their place.
 * - **"body":** Move all external stylesheet links to the end of the document.
 * - **"media":** Load stylesheets asynchronously by adding `media="not x"` and removing once loaded. <kbd>JS</kbd>
 * - **"swap":** Convert stylesheet links to preloads that swap to `rel="stylesheet"` once loaded ([details](https://www.filamentgroup.com/lab/load-css-simpler/#the-code)). <kbd>JS</kbd>
 * - **"swap-high":** Use `<link rel="alternate stylesheet preload">` and swap to `rel="stylesheet"` once loaded ([details](http://filamentgroup.github.io/loadCSS/test/new-high.html)). <kbd>JS</kbd>
 * - **"js":** Inject an asynchronous CSS loader similar to [LoadCSS](https://github.com/filamentgroup/loadCSS) and use it to load stylesheets. <kbd>JS</kbd>
 * - **"js-lazy":** Like `"js"`, but the stylesheet is disabled until fully loaded.
 * - **false:** Disables adding preload tags.
 */
type PreloadStrategy = "body" | "media" | "swap" | "swap-high" | "js" | "js-lazy";

/**
 * Controls which keyframes rules are inlined.
 *
 * - **"critical":** _(default)_ inline keyframes rules that are used by the critical CSS.
 * - **"all":** Inline all keyframes rules.
 * - **"none":** Remove all keyframes rules.
 */
type KeyframeStrategy = "critical" | "all" | "none";

/** All optional. Pass them to `new Beasties({ ... })`. */
export interface Options {
  /**
   * Base path location of the CSS files _(default: `''`)_
   */
  path?: string;
  /**
   * Public path of the CSS resources. This prefix is removed from the href _(default: `''`)_
   */
  publicPath?: string;
  /**
   * Inline styles from external stylesheets _(default: `true`)_
   */
  external?: boolean;
  /**
   * Inline external stylesheets smaller than a given size _(default: `0`)_
   */
  inlineThreshold?: number;
  /**
   * If the non-critical external stylesheet would be below this size, just inline it _(default: `0`)_
   */
  minimumExternalSize?: number;
  /**
   * Remove inlined rules from the external stylesheet _(default: `false`)_
   */
  pruneSource?: boolean;
  /**
   * Merged inlined stylesheets into a single `<style>` tag _(default: `true`)_
   */
  mergeStylesheets?: boolean;
  /**
   * additionalStylesheets Glob for matching other stylesheets to be used while looking for critical CSS.
   */
  additionalStylesheets?: string[];
  /**
   * Which {@link PreloadStrategy preload strategy} to use
   */
  preload?: PreloadStrategy | false;
  /**
   * Add `<noscript>` fallback to JS-based strategies
   */
  noscriptFallback?: boolean;
  /**
   * Inline critical font-face rules _(default: `false`)_
   */
  inlineFonts?: boolean;
  /**
   * Preloads critical fonts _(default: `true`)_
   */
  preloadFonts?: boolean;
  /**
   * Shorthand for setting `inlineFonts` + `preloadFonts`
   * - Values:
   * - `true` to inline critical font-face rules and preload the fonts
   * - `false` to don't inline any font-face rules and don't preload fonts
   */
  fonts?: boolean;
  /**
   * Controls which keyframes rules are inlined.
   */
  keyframes?: KeyframeStrategy | boolean;
  /**
   * Compress resulting critical CSS _(default: `true`)_
   */
  compress?: boolean;
  /**
   * Controls {@link LogLevel log level} of the plugin _(default: `"info"`)_
   */
  logLevel?: LogLevel;
  reduceInlineStyles?: boolean;
  /**
   * Provide a custom logger interface {@link Logger logger}
   */
  logger?: Logger;
}

export type { Logger };
