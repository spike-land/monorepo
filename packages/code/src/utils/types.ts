export interface LanguageMap {
  [key: string]: string;
}

export interface MessagePart {
  type: "text" | "code";
  content: string;
  language?: string;
  isStreaming?: boolean;
}
