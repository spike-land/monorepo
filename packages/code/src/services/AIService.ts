import { createContextManager } from "@/lib/context-manager";
import { Message, MessageContent } from "@/lib/interfaces";
import { ICode } from "@src/cSess.interface";
import { throttle } from "es-toolkit";
import { anthropicSystem, gptSystem, reminder } from "../config/aiConfig";
import { extractCodeStructure, extractCurrentTask } from "../utils/contextUtils";

export interface AIServiceConfig {
  gpt4oEndpoint: string;
  anthropicEndpoint: string;
  openAIEndpoint: string;
  updateThrottleMs: number;
  retryWithClaudeEnabled: boolean;
  setIsStreaming: (isStreaming: boolean) => void;
}

type AIEndpoint = "gpt4o" | "anthropic" | "openAI";

class StreamHandler {
  private decoder = new TextDecoder();

  async handleStream(
    reader: ReadableStreamDefaultReader<Uint8Array>,
    onUpdate: (content: string) => void,
  ): Promise<string> {
    let content = "";
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      const chunk = this.decoder.decode(value);
      content += chunk;
      onUpdate(content);
    }
    return content.trim();
  }
}

export class AIService {
  private config: AIServiceConfig;
  private streamHandler: StreamHandler;
  private cSess: ICode;
  private contextManager: ReturnType<typeof createContextManager>;

  constructor(config: AIServiceConfig, cSess: ICode, codeSpace: string) {
    this.config = config;
    this.streamHandler = new StreamHandler();
    this.cSess = cSess;
    this.contextManager = createContextManager(codeSpace);
  }

  private getEndpoint(type: AIEndpoint): string {
    const endpointMap: Record<AIEndpoint, string> = {
      gpt4o: this.config.gpt4oEndpoint,
      anthropic: this.config.anthropicEndpoint,
      openAI: this.config.openAIEndpoint,
    };
    return endpointMap[type];
  }

  private formatMessageContent(content: MessageContent): any {
    if (typeof content === "string") {
      return content;
    } else if (Array.isArray(content)) {
      return content.map(item => {
        if (item.type === "image" && item.source) {
          return {
            type: "image",
            source: {
              type: "base64",
              media_type: item.source.media_type,
              data: item.source.data,
            },
          };
        } else {
          return item;
        }
      });
    }
    return content;
  }

  private async makeAPICall(endpoint: string, messages: Message[], model = ``): Promise<Response> {
    try {
      const formattedMessages = messages.map(({ role, content }) => ({
        role,
        content: this.formatMessageContent(content),
      }));

      const response = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          stream: true,
          messages: formattedMessages,
          ...(model ? { model } : {}),
        }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return response;
    } catch (error) {
      console.error("Error making API call:", error);
      throw error;
    }
  }

  private async handleStreamingResponse(
    endpoint: string,
    messages: Message[],
    onUpdate: (content: string) => void,
    model?: string,
  ): Promise<string> {
    try {
      this.config.setIsStreaming(true);
      const response = await this.makeAPICall(endpoint, messages, model);
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error("Response body is not readable!");
      }

      const throttleUpdate = throttle(onUpdate, this.config.updateThrottleMs);
      const content = await this.streamHandler.handleStream(reader, throttleUpdate);
      onUpdate(content);

      return content;
    } catch (error) {
      console.error("Error handling streaming response:", error);
      throw error;
    } finally {
      this.config.setIsStreaming(false);
    }
  }

  async sendToAI(
    type: AIEndpoint,
    messages: Message[],
    onUpdate: (content: string) => void,
  ): Promise<Message> {
    try {
      const endpoint = this.getEndpoint(type);
      const result = await this.handleStreamingResponse(endpoint, messages, onUpdate, type === "gpt4o" ? `gpt-4o` : "");

      const lastMessage = messages[messages.length - 1];

      // Update context based on AI response
      this.contextManager.updateContext("currentTask", extractCurrentTask(result));
      this.contextManager.updateContext("codeStructure", extractCodeStructure(result));

      return {
        id: ((+lastMessage.id) + 1).toString(),
        role: "assistant",
        content: result,
      };
    } catch (error) {
      console.error("Error sending to AI:", error);
      throw error;
    } finally {
      this.config.setIsStreaming(false);
    }
  }

  async sendToGpt4o(messages: Message[], onUpdate: (content: string) => void): Promise<Message> {
    return this.sendToAI("gpt4o", messages, onUpdate);
  }

  async sendToAnthropic(messages: Message[], onUpdate: (content: string) => void): Promise<Message> {
    return this.sendToAI("anthropic", messages, onUpdate);
  }

  async continueWithOpenAI(
    fullResponse: string,
    codeNow: string,
    setMessages: React.Dispatch<React.SetStateAction<Message[]>>,
    setAICode: (code: string) => void,
  ): Promise<string> {
    const messages: Message[] = [
      { id: Date.now().toString(), role: "system", content: gptSystem },
      { id: (Date.now() + 1).toString(), role: "user", content: `${codeNow}\n**** instructions ****\n${fullResponse}` },
    ];

    const updateMessages = (newCode: string) => {
      setMessages((prevMessages) => [
        ...prevMessages,
        { id: Date.now().toString(), role: "assistant", content: newCode },
      ]);
    };

    try {
      const endpoint = this.getEndpoint("openAI");
      const response = await this.handleStreamingResponse(endpoint, messages, updateMessages);
      const modifiedCode = this.extractCodeFromResponse(response);

      if (!modifiedCode) {
        setAICode("");
        updateMessages("");
        return "";
      }

      const success = await this.cSess.setCode(modifiedCode);
      if (!success) {
        throw new Error("Error setting code");
      }
      return modifiedCode;
    } catch (error) {
      console.error("Error in AI code processing:", error);
      updateMessages(error instanceof Error ? error.message : String(error));
      throw error;
    } finally {
      this.config.setIsStreaming(false);
    }
  }

  private extractCodeFromResponse(response: string): string {
    const codeModificationRegex = /```(?:typescript?|tsx?|jsx?|javascript?)\n([\s\S]*?)```/g;
    const matches = response.match(codeModificationRegex);
    if (!matches) {
      return "";
    }
    return matches[matches.length - 1].replace(/```(?:typescript?|tsx?|jsx?|javascript?)\n|```/g, "");
  }

  async retryWithClaude(
    fullResponse: string,
    codeNow: string,
    error: unknown,
    setMessages: React.Dispatch<React.SetStateAction<Message[]>>,
    setAICode: (code: string) => void,
    prevMessages: Message[],
  ): Promise<string> {
    console.log("Retrying with Claude");
    const message: Message = {
      id: Date.now().toString(),
      role: "user",
      content: `${codeNow}\n**** instructions ****\n${fullResponse}\n**** error ****\n${
        error instanceof Error ? error.toString() : String(error)
      }\nCould you help me with this error? I'm stuck.`,
    };

    try {
      const answer = await this.sendToAnthropic([...prevMessages, message], (code) => {
        setMessages((prevMessages) => [...prevMessages, { ...message, content: code }]);
      });
      setMessages((prevMessages) => [...prevMessages, answer]);

      return await this.continueWithOpenAI(answer.content as string, codeNow, setMessages, setAICode);
    } catch (error) {
      try {
        const answer = await this.sendToGpt4o([...prevMessages, message], (code) => {
          setMessages((prevMessages) => [...prevMessages, { ...message, content: code }]);
        });
        setMessages((prevMessages) => [...prevMessages, answer]);

        return await this.continueWithOpenAI(answer.content as string, codeNow, setMessages, setAICode);
      } catch (error) {
        console.error("Error retrying with Claude:", error);
        setMessages((prevMessages) => [
          ...prevMessages,
          { id: Date.now().toString(), role: "assistant", content: "I'm sorry, I couldn't help you with this error." },
        ]);
        return "";
      }
    }
  }

  prepareClaudeContent(
    content: string,
    messages: Message[],
    codeNow: string,
    codeSpace: string,
  ): string {
    const context = this.contextManager.getFullContext();
    const contextString = JSON.stringify(context, null, 2);

    if (messages.length === 0 || codeNow !== messages[messages.length - 1]?.content) {
      return `
Current project context:
${contextString}

${
        anthropicSystem({
          fileName: codeSpace,
          fileContent: codeNow,
          userPrompt: content,
        })
      }`;
    } else {
      return `
Current project context:
${contextString}

${reminder({ userPrompt: content })}`;
    }
  }
}
