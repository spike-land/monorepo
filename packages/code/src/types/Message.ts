export type MessageContent =
  | string
  | Array<{
    type: "text" | "image";
    text?: string;
    source?: {
      type: "base64";
      media_type: "image/jpeg" | "image/png";
      data: string;
    };
  }>;

export interface Message {
  id: string;
  role: "user" | "assistant";
  content: MessageContent;
}
