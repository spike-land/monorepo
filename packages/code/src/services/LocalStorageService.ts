import { Message } from "../types/Message";

export class LocalStorageService {
  private codeSpace: string;

  constructor(codeSpace: string) {
    this.codeSpace = codeSpace;
  }

  loadMessages(): Message[] {
    return JSON.parse(
      localStorage.getItem(`chatMessages-${this.codeSpace}`) ?? "[]",
    ) as Message[];
  }

  saveAIInteraction(input: string, response: string): void {
    const interactions = JSON.parse(localStorage.getItem(`aiInteractions-${this.codeSpace}`) ?? "[]");
    interactions.push({ input, response, timestamp: Date.now() });
    localStorage.setItem(`aiInteractions-${this.codeSpace}`, JSON.stringify(interactions));
  }
}
