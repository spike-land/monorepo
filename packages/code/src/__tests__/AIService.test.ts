import { AIService } from "../services/AIService";
import { LocalStorageService } from "../services/LocalStorageService";
import { Message } from "../types/Message";

// Mock the entire LocalStorageService module
jest.mock("../services/LocalStorageService");

describe("AIService", () => {
  let aiService: AIService;
  let localStorageService: jest.Mocked<LocalStorageService>;

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Create a mocked instance of LocalStorageService
    localStorageService = {
      loadMessages: jest.fn().mockReturnValue([]),
      saveAIInteraction: jest.fn(),
    } as unknown as jest.Mocked<LocalStorageService>;

    aiService = new AIService(localStorageService);
  });

  describe("sendToAnthropic", () => {
    it("should send messages to Anthropic and return an assistant message", async () => {
      const mockMessages: Message[] = [
        { id: "1", role: "user", content: "Hello" },
      ];

      global.fetch = jest.fn().mockResolvedValue({
        ok: true,
        body: {
          getReader: () => ({
            read: jest.fn()
              .mockResolvedValueOnce({ done: false, value: new TextEncoder().encode("Assistant") })
              .mockResolvedValueOnce({ done: true }),
          }),
        },
      } as unknown as Response);

      const result = await aiService.sendToAnthropic(mockMessages);

      expect(result.role).toBe("assistant");
      expect(result.content).toBe("Assistant");
      expect(localStorageService.saveAIInteraction).toHaveBeenCalledWith("Hello", "Assistant");
    });

    it("should throw an error when the response is not OK", async () => {
      const mockMessages: Message[] = [
        { id: "1", role: "user", content: "Hello" },
      ];

      global.fetch = jest.fn().mockResolvedValue({
        ok: false,
        status: 500,
      } as Response);

      await expect(aiService.sendToAnthropic(mockMessages)).rejects.toThrow("HTTP error! status: 500");
    });
  });

  // Add more tests for other methods in AIService
});