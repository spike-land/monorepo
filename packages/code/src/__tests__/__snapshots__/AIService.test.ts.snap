// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AIService > sendToAI > should send messages to AI and return an assistant message 1`] = `
{
  "content": "Mocked assistant content",
  "id": "2",
  "role": "assistant",
}
`;
